Framework: Switch to webpack dll for vendor

blowery opened the pull request:
This PR seeks to use webpack's DLL plugin to create the `vendor` module. `vendor` holds the common pieces of the application that nearly all of the application requires, and which change fairly infrequently. 

Using the DLL plugin has two advantages over the current method of specifying an entry point. First, the build, especially rebuilds, are faster, as webpack has less code to consider when repacking the bundles. In my testing (changing code in reader/stream/post, which is used by a number of bundles), rebuilds dropped from 16s to 7s. 

Second, the hash on the vendor bundle no longer changes across deploys of calypso, unless the contents of the bundle actually changes. This means that users that use calypso often will more likely see a cache hit on the vendor bundle instead of having to reload and parse it every time we do a deploy.

To test, you'll need to spin up calypso in production mode. This should be a fail / go type test, as if anything doesn't work, calypso will not load.

Documentation refs:
http://engineering.invisionapp.com/post/optimizing-webpack/
https://robertknight.github.io/posts/webpack-dll-plugins/


Things to check out:
- [x] how does this impact the desktop build?
- [x] should the node config use the dll as well? I don't think it matters, but divergence may be painful in some way.

blowery commented:
Marking this as both needs review and in progress, as I have not tested it using a production-style docker setup as of yet.

blowery commented:
cc @aduth @gwwar

gwwar commented on client/vendor-dll.js:
```diff
@@ -0,0 +1,13 @@
+require( 'react' );
+require( 'redux' );
+require( 'store' );
+require( 'wpcom' );
```
Has our vendor bundle not been tuned in a while? I'm only seeing:

```javascript
	webpackConfig.entry.vendor = [ 'react', 'store', 'page', 'wpcom', 'jed', 'debug' ];

```

in the current webpack.config.js

blowery replied:
yeah, we haven't updated it in quite a while..

gwwar commented:
With this change we'll want to test:
[ ] - local development
[ ] - docker image  ( see docker section in PCYsg-5XR-p2 )
[ ] - wp-desktop local
[ ] - ping for help on prod wp-desktop testing

gwwar commented:
Running into the following error when attempting to build the docker image:
![screen shot 2016-09-23 at 8 42 04 am](https://cloud.githubusercontent.com/assets/1270189/18792052/eb608bf0-8169-11e6-8a9e-ab475bb850f1.png)

aduth commented on webpack-dll.config.js:
```diff
@@ -0,0 +1,58 @@
+	}
+};
+
+module.exports = webpackConfig;
```
Minor: Need to name this variable, or just `module.exports` the object directly?

blowery replied:
good catch. previous local versions changed the object based on the bundleEnv, but I pulled that out, so we can directly export the object again.

blowery replied:
fixed

aduth commented on webpack.config.js:
```diff
@@ -3,23 +3,24 @@
  * External dependencies
  */
-var webpack = require( 'webpack' ),
+let webpack = require( 'webpack' ),
```
I'd imagine most of these should be `const`, not `let`?

blowery replied:
yeah, these are all from an auto "fixing" script I run on save. I'll see about fixing them up, though they're really just cluttering up the patch.

blowery replied:
fixed em all i think

aduth commented on webpack-dll.config.js:
```diff
@@ -0,0 +1,58 @@
+	},
+	resolve: {
+		root: path.resolve( __dirname, 'client' ),
+		modulesDirectories: [ 'node_modules' ]
```
Do we need this? ([reference](https://webpack.github.io/docs/configuration.html#resolve-modulesdirectories))

blowery replied:
i _think_ we overrode it to prevent `web_modules` from being part of the search set? Likely not strictly necessary though.

blowery commented on client/reader/controller.js:
```diff
@@ -82,13 +82,13 @@ module.exports = {
 
 		if ( context.path.match( legacyPathRegexes.feedStream ) ) {
-			page.redirect( `/read/feeds/${context.params.feed_id}` );
+			page.redirect( `/read/feeds/${ context.params.feed_id }` );
```
this whole file shouldn't have been included in the patch.

blowery commented on client/vendor-dll.js:
```diff
@@ -0,0 +1,13 @@
+require( 'classnames' );
+require( 'debug' );
+require( 'i18n-calypso' );
+require( 'jed' );
```
yeah, good call.

blowery replied:
fixed

blowery commented on webpack-dll.config.js:
```diff
@@ -0,0 +1,58 @@
+const path = require( 'path' );
```
fixed

blowery commented:
> Running into the following error when attempting to build the docker image

@gwwar fixed

blowery commented:
Got this running under docker as wpcalypso. Seems to be solid so far.

gwwar commented:
@sendhil we'd really appreciate help testing this PR with Desktop

gwwar commented:
@blowery am I missing something? I'm seeing the following error when running the docker image.
<img width="1277" alt="screen shot 2016-10-04 at 3 59 07 pm" src="https://cloud.githubusercontent.com/assets/1270189/19095628/a69c5d9c-8a4b-11e6-90ea-93c8f2acb104.png">

sendhil commented:
@gwwar will take a look soon!

sendhil commented:
At the present moment the Desktop build is broken... although it's broken in master too so I'm not entirely sure if will add breakage on top. Will investigate the cause...

blowery commented:
> am I missing something? I'm seeing the following error when running the docker image.

@gwwar hmmmm looks like the minified vendor bundle didn't get built.. Is it present on disk in `public`?

gwwar commented:
@blowery verified that the docker image is working normally

sendhil commented:
So looks like Desktop broke originally over at https://github.com/Automattic/wp-calypso/pull/8421#issuecomment-251923967. Once that fix is in I can re-test this PR against the desktop app.

sendhil commented:
So this is causing Desktop to just hang at this screen:
<img width="1919" alt="screen shot 2016-10-06 at 5 15 19 pm" src="https://cloud.githubusercontent.com/assets/437043/19175065/8a0b7d9a-8be8-11e6-9d5c-35effe5d665c.png">

sendhil commented:
Some console errors:
<img width="1389" alt="screen shot 2016-10-06 at 5 16 15 pm" src="https://cloud.githubusercontent.com/assets/437043/19175083/a8a03868-8be8-11e6-8179-328d926250a1.png">

More reference on the desktop-app.js error:
<img width="804" alt="screen shot 2016-10-06 at 5 16 22 pm" src="https://cloud.githubusercontent.com/assets/437043/19175084/b56eb7cc-8be8-11e6-8f6b-66d58a119340.png">

blowery commented:
@sendhil think i fixed the desktop build. could you give it a shot?

dmsnell commented on client/vendor-dll.js:
```diff
@@ -0,0 +1,15 @@
+require( 'classnames' );
+require( 'debug' );
+require( 'i18n-calypso' );
+// lodash would go here, but we leave it out because we use tree shaking to only require the bits that we use.
```
How much of a difference are we making by tree-shaking? Do we not already use most of `lodash`?

blowery replied:
it's 4kb for the [core build](https://github.com/lodash/lodash/wiki/build-differences), which doesn't include everything we use. The full build is around 23k after gzip.

It's a good question though. We might save more by pushing all of lodash into vendor and not tree shaking at all, since we're duping some parts of lodash across chunks.

blowery replied:
That said, I'd like to land this as is. We can revisit what to stash in the vendor bundle down the line.

blowery commented:
rebased, added in a new trick!

It now extracts the manifest from the entry bundle and serves that as the new entry bundle. This keeps the hash on the build-env js file from changing when something deeper in the application changes. 

In addition, it seems like this sped up rebuilds a bit, for reals this time. Seeing ~4s rebuilds on this branch now. @dmsnell

blowery commented:
current build looks like 
```
                     media.35c28ebbae391bef56bb.js   194 kB  19, 37  [emitted]  media
                   commons.3dc0951b561f43b60a27.js   672 kB   0, 37  [emitted]
                 purchases.4a606f4142dad6db7a7f.js   378 kB   2, 37  [emitted]  purchases
               post-editor.c0575b3c195344b7c45d.js   3.1 MB   3, 37  [emitted]  post-editor
               posts-pages.c1e46b3318bd2ff6168e.js   340 kB   4, 37  [emitted]  posts-pages
reader-post-recomendations.f87a34e8e36406b9f10c.js  26.1 kB   5, 37  [emitted]  reader-post-recomendations
                  upgrades.44c025f312d8b31be214.js   803 kB   6, 37  [emitted]  upgrades
                    signup.35ab98e67384309f717f.js   621 kB   7, 37  [emitted]  signup
               reader-tags.bc47b986899a94e2a1cf.js    14 kB   8, 37  [emitted]  reader-tags
              settings-seo.5759de2710354cb486a4.js   657 kB   9, 37  [emitted]  settings-seo
                  settings.2ddf995a903dfc713573.js   522 kB  10, 37  [emitted]  settings
                   plugins.f2d8f3d04be2244e0bea.js   534 kB  11, 37  [emitted]  plugins
                     stats.031870d151cdb8aac8c1.js   358 kB  12, 37  [emitted]  stats
                     plans.a46687172b60841f7483.js   255 kB  13, 37  [emitted]  plans
                  security.e206bc3e3c2f033929ff.js   334 kB  14, 37  [emitted]  security
                     menus.29a4c3faaa590103e0bc.js   337 kB  15, 37  [emitted]  menus
                      help.3c91aa27a8db69792d37.js   230 kB  16, 37  [emitted]  help
     notification-settings.cac422c0f201835b6663.js   214 kB  17, 37  [emitted]  notification-settings
                    people.b7378459f895fd4a3c71.js   228 kB  18, 37  [emitted]  people
                    reader.1a485b00e6f21f873dc8.js   204 kB   1, 37  [emitted]  reader
                    themes.530e732368147ec7ff29.js   173 kB  20, 37  [emitted]  themes
              posts-custom.0114bb36b4d6f5dc59e8.js   184 kB  21, 37  [emitted]  posts-custom
                   sharing.ab6247f672fa0c5cb273.js   210 kB  22, 37  [emitted]  sharing
                   account.1f488870abe9f1ab958a.js   140 kB  23, 37  [emitted]  account
                     theme.cda2e78638b79e8b34d3.js   123 kB  24, 37  [emitted]  theme
          reader-following.bc7748e2eb2985923dc4.js   102 kB  25, 37  [emitted]  reader-following
             accept-invite.cac53d3fb0b1baa3eb94.js   129 kB  26, 37  [emitted]  accept-invite
              reader-start.3c60940f42e8da8ed1b5.js   130 kB  27, 37  [emitted]  reader-start
                        me.641413d3e42c40e72709.js  99.3 kB  28, 37  [emitted]  me
               reader-list.94859122a990a1ec9e08.js  66.7 kB  29, 37  [emitted]  reader-list
                       ads.c78eca651fbaa16b584d.js  68.9 kB  30, 37  [emitted]  ads
                 customize.8799b8a25c7a6efc506a.js    28 kB  31, 37  [emitted]  customize
             reader-search.0653e118c53a57524261.js  26.3 kB  32, 37  [emitted]  reader-search
         reader-activities.e8f57a04cf87a5410db3.js  5.58 kB  33, 37  [emitted]  reader-activities
             mailing-lists.58cc3a5b5d8667b2da4b.js    11 kB  34, 37  [emitted]  mailing-lists
                   paladin.dbfb8899cf0ec8fc2d53.js  3.07 kB  35, 37  [emitted]  paladin
                  discover.d0956229a680f57e5528.js  3.15 kB  36, 37  [emitted]  discover
                  manifest.3eea96a8074b6aca6b29.js  7.07 kB      37  [emitted]  manifest
          build-production.d174e8d20934864027e2.js  4.22 MB  38, 37  [emitted]  build-production
```
plus a vendor.production.js at 448K

versus master at 
```
                     media.dee57e2ffb0bbaa21a2b.js   194 kB      19  [emitted]  media
                   commons.a6d603d81792161beb77.js   672 kB       0  [emitted]
                 purchases.08ffa693abf5f5a833eb.js   378 kB       2  [emitted]  purchases
               post-editor.b1ed7092301cfc711b49.js   3.1 MB       3  [emitted]  post-editor
               posts-pages.4701d4a6e5c144fcbcf8.js   340 kB       4  [emitted]  posts-pages
reader-post-recomendations.8819e9ef096196351f04.js  26.1 kB       5  [emitted]  reader-post-recomendations
                  upgrades.ad51880a9b7d10788769.js   803 kB       6  [emitted]  upgrades
                    signup.62575d04c0f5245ec486.js   621 kB       7  [emitted]  signup
               reader-tags.0fc4f9aae8facd91ec17.js    14 kB       8  [emitted]  reader-tags
              settings-seo.036123afe97ba022d75c.js   657 kB       9  [emitted]  settings-seo
                  settings.5007c849eff547852968.js   522 kB      10  [emitted]  settings
                   plugins.19e393b4ca70b4f944ba.js   534 kB      11  [emitted]  plugins
                     stats.d70b2c1026fc4c47b088.js   358 kB      12  [emitted]  stats
                     plans.18182868bfe768f60b64.js   255 kB      13  [emitted]  plans
                  security.ae65416e74a4f0f1ef4b.js   334 kB      14  [emitted]  security
                     menus.f1fa9cab562e937e26bc.js   337 kB      15  [emitted]  menus
                      help.8d47e5a44a87654e9d2f.js   230 kB      16  [emitted]  help
     notification-settings.f05aea5cc94f0a08271d.js   214 kB      17  [emitted]  notification-settings
                    people.18c5c0d1a36d4fac6144.js   228 kB      18  [emitted]  people
                    reader.f4e16b4bec0a72d34de9.js   204 kB       1  [emitted]  reader
              posts-custom.49d2f0f0663834f6dc8b.js   184 kB      20  [emitted]  posts-custom
                   sharing.5e168b3ba9db15858cad.js   210 kB      21  [emitted]  sharing
                    themes.d04ade0623d3aaa2bfcd.js   173 kB      22  [emitted]  themes
                   account.3e4f770387785bf72590.js   140 kB      23  [emitted]  account
          reader-following.9c6ccddf006b9292fb1e.js   102 kB      24  [emitted]  reader-following
                     theme.af180c9202e1ef671c95.js   123 kB      25  [emitted]  theme
             accept-invite.d31d3f549c42bc064ba1.js   129 kB      26  [emitted]  accept-invite
              reader-start.480e846650793ed99aef.js   130 kB      27  [emitted]  reader-start
                        me.6dc148707423039a956f.js  99.3 kB      28  [emitted]  me
               reader-list.83c5f07b8f7268211ae6.js  66.7 kB      29  [emitted]  reader-list
                       ads.04b6ae31be620f8d923d.js  68.9 kB      30  [emitted]  ads
                 customize.55425db376fa1404b80a.js    28 kB      31  [emitted]  customize
             reader-search.c0768d2f606e9cbd29a5.js  26.3 kB      32  [emitted]  reader-search
         reader-activities.42e0f8f918b373a5cbe1.js  5.58 kB      33  [emitted]  reader-activities
             mailing-lists.cdaf6132ddfa5d349c0c.js    11 kB      34  [emitted]  mailing-lists
                   paladin.fba70f25ef649e4a7b78.js  3.07 kB      35  [emitted]  paladin
                  discover.b8ef9cdde7b743ffa32f.js  3.15 kB      36  [emitted]  discover
                    vendor.a6d603d81792161beb77.js   459 kB      37  [emitted]  vendor
          build-production.a6d603d81792161beb77.js  4.94 MB      38  [emitted]  build-production
```

blowery commented:
@gwwar @aduth any chance you can try this out? I'd like to get it landed.

aduth commented on Makefile:
```diff
@@ -149,6 +149,10 @@ server/devdocs/search-index.js: $(MD_FILES) $(ALL_DEVDOCS_JS)
 server/devdocs/components-usage-stats.json: $(COMPONENTS_USAGE_STATS_FILES) $(COMPONENTS_USAGE_STATS_JS)
 	@$(COMPONENTS_USAGE_STATS_JS) $(COMPONENTS_USAGE_STATS_FILES)
 
+build-dll: .env node_modules
```
Why `.env` ?

blowery replied:
don't recall. Doesn't seem like it's actually needed.

aduth commented on webpack-dll.config.js:
```diff
@@ -0,0 +1,62 @@
+	resolve: {
+		root: path.resolve( __dirname, 'client' )
+	}
+};
```
At some point we should consider extracting out common bits across all our Webpack configs and having individual ones extend via e.g. `_.merge`.

blowery replied:
At some point, yes, we should. It's gotten a bit nuts.

aduth commented on server/pages/index.js:
```diff
@@ -80,7 +81,11 @@ function generateStaticUrls( request ) {
 
-	assets = request.app.get( 'assets' );
+	// vendor dll
+	urls.vendor = urls[ 'vendor.' + config( 'env' ) + '.js' ];
```
Any reason not to do the lookup by env in the Jade template instead of aliasing it here? We're already doing similar for `jsFile`, e.g.

```jade
script(src=urls[ 'vendor.' + env + '.js' ])
```

blowery replied:
don't recall. checking. likely just an oversight.

blowery replied:
ah, different env! the env in the jade template is the CALYPSO_ENV while this env is the NODE_ENV. We only need two vendor bundles, one for prod and one for dev, not one for each CALYPSO_ENV.

aduth commented on server/pages/index.js:
```diff
@@ -22,28 +22,29 @@ import { createReduxStore } from 'state';
 	CALYPSO_ENV = process.env.CALYPSO_ENV || process.env.NODE_ENV || 'development';
 
-var staticFiles = [
+const staticFiles = [
```
Unrelated, but many of these `staticFiles` are large and their contents are loaded synchronously during the hash step for `getUrl`. Would be nice to have some parallelization there.

aduth commented on webpack-dll.config.js:
```diff
@@ -0,0 +1,62 @@
+
+module.exports = {
+	entry: {
+		vendor: [ path.join( __dirname, 'client', 'vendor-dll.js' ) ]
```
Couldn't we pass all the vendor libs directly in the array here, instead of as a separate file in `client`? Basing this assumption [on the current configuration](https://github.com/Automattic/wp-calypso/blob/f285c47e26d3aa3e9093a0d8f98b4b84ddc39cb0/webpack.config.js#L98).

blowery replied:
inlined it in https://github.com/Automattic/wp-calypso/pull/8161/commits/b544c2fa0e396b2ae95e81ac4653602abf34dd12

gwwar commented:
👍 Retested this in local dev, docker, and Desktop dev environments. I didn't run into any build or js  errors.

blowery commented:
@aduth any other comments on this one?

aduth commented:
Sorry for the delay, wanted to give this a go in local production environment. LGTM 👍

dmsnell commented:
so happy thanks @blowery and all reviewers!

mtias commented:
Thanks for pushing this one through!
